internal/procedures/users/upsert.go

```diff
@@ -142,6 +149,11 @@ func insertUser(user *model.ServerUser, options UpsertUsersOptions, rawData *mod
 		result.Enrolled = append(result.Enrolled, courseID)
 	}
 
+	err = user.Validate()
+	if err != nil {
+		return model.NewUserOpResultValidationError("-1023", user.Email, err)
```

Comment:
Let's get a test case for this.
I imagine you can probably just insert the user, but with a bad email or something.

Reply:
This error should never be returned. We perform a weak validation and return error code -1003 for all common validation errors. Since this validation check is the strong version, it could only catch unknown server roles (that wouldn't be caught be error -1003).

But, we explicitly set the role to be a known value before this validation.

I included this validation as a sanity check.

Reply:
kk, sounds good.